desktopToDarwinBundle: Patch Exec field codes

toonn opened the pull request:
###### Description of changes

Desktop entry files can include field codes, e.g., `%f`, `%U`, in the `Exec=` entry. macOS doesn't understand these so we drop them here and add the ability to override the executable entirely.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

toonn committed:
desktopToDarwinBundle: Fixup Exec

The "Exec" key in desktop items sometimes has one of the `%f`, `%F`,
`%u` and `%U` suffixes, which specify whether the command takes a file,
multiple files or a generalized URL or URLs. Darwin application bundles
do no understand this syntax so we do the next best thing, which is
simply dropping it.

toonn committed:
desktopToDarwinBundle: Complete field code removal

Checked the desktop entry spec, there's other field codes than `%[fFuU]`
and those can in fact occur more than once, hence dropping '$' and
adding `/g`.

toonn committed:
desktopToDarwinBundle: Add X-macOS-Exec and log editing Exec

Co-authored-by: milahu <milahu@gmail.com>

hexagonal-sun commented on pkgs/build-support/setup-hooks/desktop-to-darwin-bundle.sh:
```diff
@@ -150,7 +150,16 @@ convertDesktopFile() {
+      local -r execRaw=$(getDesktopParam "${file}" "Exec")
+      local -r exec="${execRaw// %[fFuUick]}"
+      if [[ "$exec" != "$execRaw" ]]; then
+        echo "desktopToDarwinBundle: Application bundles do not understand desktop entry field codes. Changed '$execRaw' to '$exec'."
```
Should we consider `exit 1` here to stop the build?  My concern is that this message may get lost and then we would be executing something that isn't intended.

hexagonal-sun replied:
Agreed.  To be fair, I hadn't looked up the spec on what these flags mean.  Having read through then now, I don't think they're applicable to MacOS.

reckenrode commented:
I did a grep of nixpkgs to see if there were any apps that depend on `%c`, `%i`, or `%k` and found that KMPlayer depends on both `%c` and `%i`. I have no idea whether it actually works on Darwin. It’s possible upstream packages are shipping desktop files that depend on these flags, but I don’t know of a good way to find that out (maybe someone with a desktop NixOS install could grep the store?).

reckenrode commented on pkgs/build-support/setup-hooks/desktop-to-darwin-bundle.sh:
```diff
@@ -150,7 +150,16 @@ convertDesktopFile() {
     local -r sharePath=$(dirname "$(dirname "$file")")
     local -r name=$(getDesktopParam "${file}" "^Name")
-    local -r exec=$(getDesktopParam "${file}" "Exec")
+    local -r macOSExec=$(getDesktopParam "${file}" "X-macOS-Exec")
+    if [[ "$macOSExec" ]]; then
+      local -r exec="$macOSExec"
+    else
+      local -r execRaw=$(getDesktopParam "${file}" "Exec")
+      local -r exec="${execRaw// %[fFuUick]}"
+      if [[ "$exec" != "$execRaw" ]]; then
+        echo "desktopToDarwinBundle: Application bundles do not understand desktop entry field codes. Changed '$execRaw' to '$exec'."
+      fi
+    fi
```
If you `open Foo.app --args <args>`, the `<args>` will passed to the executable. I had to do that at work with the Microsoft Teams app to disable GPU rendering to save battery (before it got that in the app’s preferences). I also have a script that to open a file in Excel (handy for `.csv` files when you don’t want to open them in Numbers but don’t want to make Excel the default app).

However, like I said before, I’m not sure whether those matter or if translating them is the right thing to do on macOS. According to the spec, `%f` and `%u` are used when opening a file or URL. macOS doesn’t seem to care. When I hacked element-desktop to support for the `element:` URL scheme (to use GitHub to authenticate), I don’t think I bothered to change `%f` to `$1`, but I can’t remember. Seems worth revisiting once support for file associations and URL schemes is added.

Setting those aside, it’s the other three (`%i`, `%c`, and `%k`) I’m more concerned about. Those are static (referring back to information in the .desktop file or the .desktop file itself). My assumption is if an app specifies those in its .desktop file, it expects to get that information on the command-line. See my other comment re: apps that use these flags. There is one in nixpkgs I identified (though it is marked as broken on Darwin). I did a search of GitHub, and it seems to be used by Qt apps.

toonn replied:
I concur that it's better to actually implement the field codes we can.
DEs are supposed to do something smart like spawn a process for every selected file when launching a selection using a desktop entry with `%f`/`%u`. Should we emulate this in the wrapper script? Rather than simply using `$1` we'd have to fork for every argument passed or something.

toonn replied:
Just pushed some commits handling each of the field codes. (And rebased on master while I was at it.)

I haven't added any special handling of `%f` or `%u`, so multiple arguments passed to such a program would simply silently be ignored. I'm not sure that's very nice behavior. Should we loop over `$@` and fork the program for each argument?

reckenrode replied:
> I haven't added any special handling of `%f` or `%u`, so multiple arguments passed to such a program would simply silently be ignored. I'm not sure that's very nice behavior. Should we loop over `$@` and fork the program for each argument?

This is the part I’ve not been sure about, so I did some digging.

First, I put together a simple app that just displays the command-line arguments and associated it with the `public.text` UTI. When I opened text files, it displayed just the `argv[0]`. Not sure whether that was because the app was a SwiftUI app (and the framework was mangling `CommandLine.arguments`), I also tried an Objective-C app that just displays an `NSAlert`. Same thing.

So I did some searching to see how apps are supposed to receive files and URLs. The [answer][so] is Finder calls one of the following methods on the app’s `NSApplicationDelegate`. I don’t know how the invocation works, but if it uses the command-line as part of the process, those arguments do not make it to `main`.

* [`application:openFile`][openFile]
* [`application:openFiles`][openFiles]
* [`application:openURLs`][openURLs]

I think the only approach that will work is to strip `%f` and `%u` and rely on whatever UI framework the app uses to implement the necessary support in its platform-specific stuff for Darwin. Both [GKT][gtk-delegate] and [Qt][qt-delegate] do this, and Electron presumably does as well since my association with the `element:` URL scheme worked without handling it in the script.

[so]: https://stackoverflow.com/questions/2998203/how-to-make-finder-open-with-work-for-my-application-xcode-os-x
[openFile]: https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428612-application
[openFiles]: https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428742-application
[openURLs]: https://developer.apple.com/documentation/appkit/nsapplicationdelegate/2887193-application
[gtk-delegate]: https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/gtkapplication-quartz.c
[qt-delegate]: https://github.com/qt/qtbase/blob/3fcdb6cb6e35a37f9b511ec2705336102c194d6b/src/plugins/platforms/cocoa/qcocoaapplicationdelegate.mm

toonn replied:
Does this mean I should drop the changes for `%[fFuU]` and strip them out like before? Or are `$1` and `$@` still fine?

reckenrode replied:
Well, I doubt it would hurt to keep those changes. If someone `open`ed the app, the file or URL would get passed to it. I’ll go ahead and improve. Thanks for the changes!

toonn committed:
desktopToDarwinBundle: Implement %k Exec field code

toonn committed:
desktopToDarwinBundle: Change empty directory test

`ls -1 "$iconsdir/"*` listed the source directory for me when the glob
had no matches. Switching to `-A` circumvents this problem and has the
added advantage that it cannot run into argument list length limits.

toonn committed:
desktopToDarwinBundle: Implement %c Exec field code

toonn committed:
desktopToDarwinBundle: Implement %i Exec field code

toonn committed:
desktopToDarwinBundle: Implement %f and %u Exec field codes

`%f` and `%u` are used to signal the program only accepts a single file
or URI argument. I do not believe there's a way to signal this
information to macOS but it is possible the program really won't work if
multiple files are passed and it's possible the relative position of
`%i`, `%c` or `%k` matters. So we replace `%f` or `%u` with `$1`. That
way we only pass one file in the (possibly significant) position of the
field code.

toonn committed:
desktopToDarwinBundle: Implement %F and %U Exec field codes

Similar to the implementation of the `%f` and `%u` field codes. In this
case the amount of arguments passed poses no problem but the position
could, at least in theory.

This finishes the implementation of all the non-deprecated field codes.
As a part of that, repetitions of field codes are left alone. Originally
all field codes were removed. Now we replace only the first occurence.
This is correct for at least `%f`, `%u`, `%F` and `%U` because at most
one of them is permitted.

Shortcomings:

  1. We replace `%[cfFikuU]` patterns one at a time. This means if the
     right field code appears as part of the rest of the `Exec` field or
     in a field code that was substituted earlier.

  2. If any field code is repeated, only the first occurence is
     substituted.
